modules/upscaler_utils.py

```diff
@@ -1,50 +1,27 @@
 import logging
 from typing import Callable
 
-import numpy as np
 import torch
 import tqdm
 from PIL import Image
+import torchvision.transforms.v2 as T
 
 from modules import images, shared, torch_utils
 
 logger = logging.getLogger(__name__)
 
 
-def pil_image_to_torch_bgr(img: Image.Image) -> torch.Tensor:
-    img = np.array(img.convert("RGB"))
-    img = img[:, :, ::-1]  # flip RGB to BGR
-    img = np.transpose(img, (2, 0, 1))  # HWC to CHW
-    img = np.ascontiguousarray(img) / 255  # Rescale to [0, 1]
-    return torch.from_numpy(img)
-
-
-def torch_bgr_to_pil_image(tensor: torch.Tensor) -> Image.Image:
-    if tensor.ndim == 4:
-        # If we're given a tensor with a batch dimension, squeeze it out
-        # (but only if it's a batch of size 1).
-        if tensor.shape[0] != 1:
-            raise ValueError(f"{tensor.shape} does not describe a BCHW tensor")
-        tensor = tensor.squeeze(0)
-    assert tensor.ndim == 3, f"{tensor.shape} does not describe a CHW tensor"
-    # TODO: is `tensor.float().cpu()...numpy()` the most efficient idiom?
-    arr = tensor.float().cpu().clamp_(0, 1).numpy()  # clamp
-    arr = 255.0 * np.moveaxis(arr, 0, 2)  # CHW to HWC, rescale
-    arr = arr.round().astype(np.uint8)
-    arr = arr[:, :, ::-1]  # flip BGR to RGB
-    return Image.fromarray(arr, "RGB")
```

Comment:
I'd like it if these function names were retained, even if they did just call the Torchvision transforms. The assertions and comments are there for a reason too :)

Reply:
For the time being I've just removed the Torchvision transforms. I noticed when I was testing a DAT model, there was a large color shift. I'm unclear if this is just because I had added FP32 conversions and clamps than your original code or something else entirely. I should probably check against ChaNNier later to see which was the closest match.

It would appear compared to your code, the Torchvision functions I used did the following internally, which I then clamped every time after converting to float32 Tensors before converting to PIL Image:
```
ToPILImage
...
pic = pic.permute((1, 2, 0))
pic = pic.numpy(force=True)
npimg = pic
if np.issubdtype(npimg.dtype, np.floating) and mode != "F":
    npimg = (npimg * 255).astype(np.uint8)
if mode is None and npimg.dtype == np.uint8:
    mode = "RGB"
return Image.fromarray(npimg, mode=mode)
...
```
```
PILToTensor
...
img = torch.as_tensor(np.array(pic, copy=True))
img = img.view(pic.size[1], pic.size[0], F_pil.get_image_num_channels(pic))
img = img.permute((2, 0, 1))
return img
...
```
```
ToDtype
...
if float_input:
    # float to float
    if float_output:
        return image.to(dtype)
    # float to int    
    eps = 1e-3
    max_value = float(_max_value(dtype))
    return image.mul(max_value + 1.0 - eps).to(dtype)
else:
    # int to float
    if float_output:
        return image.to(dtype).mul_(1.0 / _max_value(image.dtype))
    # int to int
    num_value_bits_input = _num_value_bits(image.dtype)
    num_value_bits_output = _num_value_bits(dtype)

    if num_value_bits_input > num_value_bits_output:
        return image.bitwise_right_shift(num_value_bits_input - num_value_bits_output).to(dtype)
    else:
        return image.to(dtype).bitwise_left_shift_(num_value_bits_output - num_value_bits_input)
...
```